pkg/networkservice/core/trace/common.go

```diff
@@ -17,57 +17,156 @@
 	"github.com/networkservicemesh/sdk/pkg/tools/spanhelper"
 )
 
 func logRequest(span spanhelper.SpanHelper, request proto.Message) {
-	connInfo, ok := tracehelper.FromContext(span.Context())
+	connInfo, ok := trace(span.Context())
 	if ok && !proto.Equal(connInfo.Request, request) {
-		requestDiff := diff(connInfo.Request, request)
-		span.LogObject("request diff", requestDiff.Interface())
+		if connInfo.Request != nil {
+			requestDiff, hadChanges := Diff(connInfo.Request.ProtoReflect(), request.ProtoReflect())
+			if hadChanges {
+				span.LogObject("request-diff", requestDiff)
+			}
+		} else {
+			span.LogObject("request", request)
+		}
 		connInfo.Request = proto.Clone(request)
 	}
 }
 
 func logResponse(span spanhelper.SpanHelper, response proto.Message) {
-	connInfo, ok := tracehelper.FromContext(span.Context())
+	connInfo, ok := trace(span.Context())
 	if ok && !proto.Equal(connInfo.Response, response) {
-		responseDiff := diff(connInfo.Response, response)
-		span.LogObject("response diff", responseDiff.Interface())
+		if connInfo.Response != nil {
+			responseDiff, changed := Diff(connInfo.Response.ProtoReflect(), response.ProtoReflect())
+			if changed {
+				span.LogObject("response-diff", responseDiff)
+			}
+		} else {
+			span.LogObject("response", response)
+		}
 		connInfo.Response = proto.Clone(response)
 		return
 	}
-	span.LogValue("response", "")
 }
 
-func diff(oldMessage, newMessage proto.Message) protoreflect.Message {
+// Diff - calculate a protobuf messge diff
+func Diff(oldMessage, newMessage protoreflect.Message) (map[string]interface{}, bool) {
 	if oldMessage == nil || reflect.ValueOf(oldMessage).IsNil() {
-		return newMessage.ProtoReflect()
+		return nil, true
 	}
 
-	diffMessage := oldMessage.ProtoReflect().New()
-	oldReflectMessage := oldMessage.ProtoReflect()
-
-	newMessage.ProtoReflect().Range(func(descriptor protoreflect.FieldDescriptor, newValue protoreflect.Value) bool {
-		oldValue := oldReflectMessage.Get(descriptor)
-		if !reflect.DeepEqual(oldValue, newValue) {
-			if fieldMessage, ok := newValue.Interface().(protoreflect.Message); ok {
-				fieldDiff := diff(
-					oldValue.Interface().(protoreflect.Message).Interface(),
-					fieldMessage.Interface(),
-				)
-				newValue = protoreflect.ValueOf(fieldDiff)
+	diffMessage := map[string]interface{}{}
+	oldReflectMessage := oldMessage
+
+	// Marker we had any changes
+	changes := 0
+
+	newMessage.Range(func(descriptor protoreflect.FieldDescriptor, newRefValue protoreflect.Value) bool {
+		rawOldValue := oldReflectMessage.Get(descriptor)
+		oldValue := rawOldValue.Interface()
+		newValue := newRefValue.Interface()
+
+		if descriptor.Cardinality() == protoreflect.Repeated {
+			if newList, ok := newValue.(protoreflect.List); ok {
+				oldList := oldValue.(protoreflect.List)
+
+				originMap := map[string]protoreflect.Value{}
+				for i := 0; i < oldList.Len(); i++ {
+					originMap[fmt.Sprintf("%d", i)] = oldList.Get(i)
```

Comment:
strconv.Itoa works some effective then fmt.Sprint. https://gist.github.com/evalphobia/caee1602969a640a4530
```
BenchmarkItoa                           30000000           52.6 ns/op         2 B/op           1 allocs/op
BenchmarkItoaSprint                     5000000            254 ns/op          16 B/op          2 allocs/op

```
```suggestion
					originMap[strconv.Itoa(i)] = oldList.Get(i)
```